bundles/org.openhab.voice.pipertts/src/main/java/org/openhab/voice/pipertts/internal/PiperTTSService.java

```diff
@@ -0,0 +1,417 @@
+                    logger.debug("Using preloaded voice model");
+                    preloadedModel.consumers.incrementAndGet();
+                    voiceModel = preloadedModel;
+                    usingPreloadedModel = true;
+                } else {
+                    unloadModel();
+                    logger.debug("Loading voice model...");
+                    voiceModel = loadModel(ttsVoice);
+                    synchronized (this) {
+                        usingPreloadedModel = voiceModel.equals(this.preloadedModel);
+                    }
+                }
+            } catch (IOException e) {
+                throw new TTSException("Unable to load voice model: " + e.getMessage());
+            }
+            try {
+                logger.debug("Generating audio for: '{}'", text);
+                buffer = getPiper().textToAudio(voiceModel.piperVoice, text);
+                logger.debug("Generated {} samples of audio", buffer.length);
+            } catch (IOException e) {
+                throw new TTSException("Voice generation failed: " + e.getMessage());
+            }
+        } catch (PiperJNI.NotInitialized | LibraryNotLoaded e) {
+            throw new TTSException("Piper not initialized, try restarting the add-on.");
+        } catch (RuntimeException e) {
+            logger.warn("RuntimeException running text to audio: {}", e.getMessage());
+            throw new TTSException("There was an error running Piper");
+        } finally {
+            if (voiceModel != null) {
+                if (!usingPreloadedModel
+                        || voiceModel.consumers.decrementAndGet() == 0 && !voiceModel.equals(this.preloadedModel)) {
+                    logger.debug("Unloading voice model");
+                    voiceModel.close();
+                } else {
+                    logger.debug("Skipping voice model unload");
+                }
+            }
+        }
+        try {
+            logger.debug("Return re-encoded audio stream");
+            return getAudioStream(buffer, voiceModel.sampleRate, audioFormat);
+        } catch (IOException e) {
+            throw new TTSException("Error while creating audio stream: " + e.getMessage());
+        }
+    }
+
+    private VoiceModel loadModel(PiperTTSVoice voice) throws IOException, PiperJNI.NotInitialized, LibraryNotLoaded {
+        if (!Files.exists(voice.voiceModelPath()) || !Files.exists(voice.voiceModelConfigPath())) {
+            throw new IOException("Missing voice files");
+        }
+        PiperJNI piper = getPiper();
+        PiperVoice piperVoice;
+        VoiceModel voiceModel;
+        piperVoice = piper.loadVoice(voice.voiceModelPath(), voice.voiceModelConfigPath(), voice.speakerId.orElse(-1L));
+        voiceModel = new VoiceModel(voice, piperVoice, piperVoice.getSampleRate(), new AtomicInteger(1), logger);
+        if (config.preloadModel) {
+            synchronized (this) {
```

Comment:
It is better to create a private lock object and synchronize on that. You never know what other code is also synchronizing on the instance. This prevents unexpected deadlocks.

Reply:
I have changed the synchronized blocks to using a ReentrantLock instance. Is that the correct way?

Reply:
Since it only does simple locking/unlocking you can simplify it using:

```java
private Object modelLock = new Object();
```

Then you can just use: 

```java
synchronized (modelLock) {
   ....
}
```

That way there will be less code and locking mistakes. :slightly_smiling_face:

Reply:
Done. Thank you for the clarification and the rest of the review!